PE-75 kube autoscaler per ASG

sukenshah opened the pull request:
## Overview
The current version of `kube-aws-autoscaler` only allows fine-tuning parameters at the global level, which means all the worker pools are scaled up and down using the same criteria. This is good for a small number of similar profile worker pools but restricts us from setting up different scaling profiles/policies per worker pool. For instance, the scaling and buffer for interactive worker pools will be really different from the batch and core worker pools.

## Changes
The proposed solution is to fork the kube-aws-autoscaler repo and implement a new strategy where the autoscaler will use the following parameters from the tags of ASG of worker pools if present:

1. buffer memory %
2. buffer cpu %
3. buffer spare node

If the ASG doesn't have the tags then the autoscaler will use the global configured params, if they are missing then it'll fallback to hardcoded defaults.

## Unit testing
Added following unit tests for `calculate_buffer_per_auto_scaling_group` method:

test_calculate_buffer_per_auto_scaling_group_with_no_asg_tag
test_calculate_buffer_per_auto_scaling_group_with_cpu_override
test_calculate_buffer_per_auto_scaling_group_with_memory_override
test_calculate_buffer_per_auto_scaling_group_with_pods_override
test_calculate_buffer_per_auto_scaling_group_with_nodes_override
test_calculate_buffer_per_auto_scaling_group_with_all_override
test_calculate_buffer_per_auto_scaling_group_with_multiple_groups_and_partial_overrides

Some tests specifically setup using the buffer per asg for `calculate_required_auto_scaling_group_sizes` method:

test_calculate_required_auto_scaling_group_sizes_with_asg_overrides_scaleup_no_spare_node
test_calculate_required_auto_scaling_group_sizes_with_asg_overrides_scaleup_with_spare_node
test_calculate_required_auto_scaling_group_sizes_with_asg_overrides_no_scale
test_calculate_required_auto_scaling_group_sizes_with_asg_overrides_scale_down


## Integrated testing
Tested the changes by running `kube-aws-autoscaler` locally against the staging env in debug mode and with all fixed buffers set to zero and % buffer set to 10.

### 1. Core worker pool:
**ASG Settings**

Tag | Value
-- | --
kube-aws-autoscaler:buffer-cpu-percentage | 0
kube-aws-autoscaler:buffer-memory-percentage | 5
kube-aws-autoscaler:buffer-pods-percentage | 0
kube-aws-autoscaler:buffer-spare-nodes | 0

<img width="1336" alt="kube-aws-autoscaler-all-overrides" src="https://user-images.githubusercontent.com/43122428/47772038-817b0e80-dd39-11e8-80fd-50eee4d0fd1e.png">

### 2.  Worker pool A:
**ASG Settings**

Tag | Value
-- | --
kube-aws-autoscaler:buffer-memory-percentage | 5

<img width="1332" alt="kube-aws-autoscaler-partial-overrides" src="https://user-images.githubusercontent.com/43122428/47772090-b2f3da00-dd39-11e8-9079-228bcdaab046.png">

Please note in this case the rest of the params are defaulted to 10%

Commit:
PE-75 Updated the kube autoscaler to override the global config params with ASG ones if specified as part of ASG tags.

Commit:
PE-75 Added deployment descriptor.

Commit:
Added create ecr repo step in build pipeline.

Commit:
Checking if tags are present in the describe autoscaling group response.

Commit:
Aligning tests with recent code changes.

Commit:
Added pytest package to requirements.

Commit:
Fixing issue with already broken test_get_nodes.

Commit:
Using python 3.7 version for the build and test.

ronny commented on kube_aws_autoscaler/main.py:
```diff
@@ -231,8 +232,42 @@ def slow_down_downscale(asg_sizes: dict, nodes_by_asg_zone: dict, scale_down_ste
+    if len(groups) > 0:
+        response = autoscaling.describe_auto_scaling_groups(AutoScalingGroupNames=groups)
+        for asg in response['AutoScalingGroups']:
+            if 'Tags' in asg.keys() and len(asg['Tags']) > 0:
```
```suggestion
            if 'Tags' in asg and asg['Tags']:
```

mousavian replied:
WOW @ronny . TIL `suggestion`

ronny commented on kube_aws_autoscaler/main.py:
```diff
@@ -231,8 +232,42 @@ def slow_down_downscale(asg_sizes: dict, nodes_by_asg_zone: dict, scale_down_ste
+                buffer_info = {}
+                for asg_tag in asg['Tags']:
+                    tag_key = asg_tag['Key']
+                    if tag_key in ASG_ALLOWED_TAGS.keys():
```
```suggestion
                    if tag_key in ASG_ALLOWED_TAGS:
```

sukenshah replied:
Fixed.

ronny commented on kube_aws_autoscaler/main.py:
```diff
@@ -231,8 +232,42 @@ def slow_down_downscale(asg_sizes: dict, nodes_by_asg_zone: dict, scale_down_ste
+                    if tag_key in ASG_ALLOWED_TAGS.keys():
+                        buffer_info[ASG_ALLOWED_TAGS[tag_key]] = int(asg_tag['Value'])
+
+                if len(buffer_info.keys()) > 0:
```
```suggestion
                if buffer_info:
```

```
>>> bool({'a':1})
True
>>> bool({})
False
```

sukenshah replied:
Fixed.

ronny commented on kube_aws_autoscaler/main.py:
```diff
@@ -46,6 +46,7 @@
 
 logger = logging.getLogger('autoscaler')
 
+ASG_ALLOWED_TAGS = {'buffer-memory-percentage': 'memory', 'buffer-cpu-percentage': 'cpu', 'buffer-pods-percentage': 'pods', 'buffer-spare-nodes': 'nodes'}
```
maybe worth splitting these into lines to make it easier to read

sukenshah replied:
Fixed.

msessa commented on kube_aws_autoscaler/main.py:
```diff
@@ -46,6 +46,7 @@
 
 logger = logging.getLogger('autoscaler')
 
+ASG_ALLOWED_TAGS = {'buffer-memory-percentage': 'memory', 'buffer-cpu-percentage': 'cpu', 'buffer-pods-percentage': 'pods', 'buffer-spare-nodes': 'nodes'}
```
Perhaps we should use a prefix on those tags to help identify their purpose at a glance.
Ex:
  aws-autoscaler-buffer-memory-percentage
  aws-autoscaler-buffer-cpu-percentage
  etc

sukenshah replied:
Added `kube-aws-autoscaler:` as prefix.

Commit:
Removing deployment descriptor.

Commit:
Fixing code review comments.

Commit:
Updated the ASG tag to add kube-aws-autoscaler: as prefix.

Commit:
Added following tests for the calculate_buffer_per_auto_scaling_group

test_calculate_buffer_per_auto_scaling_group_with_no_asg_tag
test_calculate_buffer_per_auto_scaling_group_with_cpu_override
test_calculate_buffer_per_auto_scaling_group_with_memory_override
test_calculate_buffer_per_auto_scaling_group_with_pods_override
test_calculate_buffer_per_auto_scaling_group_with_nodes_override
test_calculate_buffer_per_auto_scaling_group_with_all_override
test_calculate_buffer_per_auto_scaling_group_with_multiple_groups_and_partial_overrides

ronny commented on kube_aws_autoscaler/main.py:
```diff
@@ -231,8 +237,42 @@ def slow_down_downscale(asg_sizes: dict, nodes_by_asg_zone: dict, scale_down_ste
+        asg_name, zone = key
+        groups.append(asg_name)
+
+    if len(groups) > 0:
```
```suggestion
    if groups:
```

ronny commented on kube_aws_autoscaler/main.py:
```diff
@@ -231,8 +237,42 @@ def slow_down_downscale(asg_sizes: dict, nodes_by_asg_zone: dict, scale_down_ste
+                    if tag_key in ASG_ALLOWED_TAGS:
+                        buffer_info[ASG_ALLOWED_TAGS[tag_key]] = int(asg_tag['Value'])
+
+                if buffer_info.keys():
```
```suggestion
                if buffer_info:
```

ronny commented on kube_aws_autoscaler/main.py:
```diff
@@ -231,8 +237,42 @@ def slow_down_downscale(asg_sizes: dict, nodes_by_asg_zone: dict, scale_down_ste
+
+                if buffer_info.keys():
+                    # Fill the missing buffer resource with global
+                    if 'nodes' not in buffer_info.keys():
```
```suggestion
                    if 'nodes' not in buffer_info:
```

ronny commented on kube_aws_autoscaler/main.py:
```diff
@@ -231,8 +237,42 @@ def slow_down_downscale(asg_sizes: dict, nodes_by_asg_zone: dict, scale_down_ste
+                    if 'nodes' not in buffer_info.keys():
+                        buffer_info['nodes'] = buffer_spare_nodes
+                    for buffer_resource in RESOURCES:
+                        if buffer_resource not in buffer_info.keys():
```
```suggestion
                        if buffer_resource not in buffer_info:
```

ronny commented on kube_aws_autoscaler/main.py:
```diff
@@ -246,7 +286,15 @@ def calculate_required_auto_scaling_group_sizes(nodes_by_asg_zone: dict, usage_b
                 requested[resource] += val
-        requested_with_buffer = apply_buffer(requested, buffer_percentage, buffer_fixed)
+
+        if asg_name in buffer_per_asg.keys():
```
```suggestion
        if asg_name in buffer_per_asg:
```

sukenshah commented:
Hello @ronny  @mousavian and @msessa I have added some unit tests for `calculate_buffer_per_auto_scaling_group `. See the PR description.

ronny reviewed (approved):
Tests are 💯. Awesome work!

sukenshah commented:
Added some more following tests to cover ASG desired size.

test_calculate_required_auto_scaling_group_sizes_with_asg_overrides_scaleup_no_spare_node
test_calculate_required_auto_scaling_group_sizes_with_asg_overrides_scaleup_with_spare_node
test_calculate_required_auto_scaling_group_sizes_with_asg_overrides_no_scale
test_calculate_required_auto_scaling_group_sizes_with_asg_overrides_scale_down

Commit:
Some tests specifically setup using the buffer per asg for calculate_required_auto_scaling_group_sizes method:

test_calculate_required_auto_scaling_group_sizes_with_asg_overrides_scaleup_no_spare_node
test_calculate_required_auto_scaling_group_sizes_with_asg_overrides_scaleup_with_spare_node
test_calculate_required_auto_scaling_group_sizes_with_asg_overrides_no_scale
test_calculate_required_auto_scaling_group_sizes_with_asg_overrides_scale_down

mousavian commented on kube_aws_autoscaler/main.py:
```diff
@@ -231,8 +237,42 @@ def slow_down_downscale(asg_sizes: dict, nodes_by_asg_zone: dict, scale_down_ste
+                    if tag_key in ASG_ALLOWED_TAGS:
+                        buffer_info[ASG_ALLOWED_TAGS[tag_key]] = int(asg_tag['Value'])
+
+                if buffer_info.keys():
```
this if statement `if buffer_info.keys():` is not necessary as you check again in lines 261 & 264 as well.

sukenshah replied:
The reason for checking this is to make sure we only add buffer info for an ASG with valid tags. It'll avoid creating buffer_info for ASG with tags but none with `kube-aws-autoscaler:`

mousavian reviewed (approved):
great job @sukenshah

Commit:
Minor refactoring.

Commit:
Updated readme.

sukenshah committed:
Updated readme.
